[WIP]Updated installation instructions

simon66 opened the pull request:
While installing Thimble, I ran into some installation problems. I didn't know if Bramble was required or not (Since it only mentions needing it if you plan to install it locally).
I've updated the instructions. Hopefully is more clear :)
I've removed the Bramble installation part since the Bramble repository has a really good, in-depth installation instruction. 
Is there anything you would like me to add/remove?

Pomax commented:
Hi @simon66, thanks for the PR, but I'd like to suggest going in the opposite direction =)

While it may seem a good idea to "remove extraenous information" from a README, for open source projects it is usually actually far better to have as much information in that file. In this case, that means that rather than pruning information about Bramble, these instructions should be preserved, but moved into a section where that information makes sense. To that end, we should make it clear that there are two ways to install things:

1. Vagrant, a one-stop way to install everything, and
2. When vagrant won't work for you, the proper full list of which projects need to be installed in parallel, which includes thimble, bramble, id, login, and publish.

Rather than taking everything except for Vagrant out of the README.md, it will be far more valuable to split the README instructions up into two sections: an initial part "using vagrant" that explains things in terms of just using Vagrant, and then a subsequent more elaborate "manual installation" part that explains how to install the components manually, should you need to.

And you might need to go through the history of the README.md a bit because when the vagrant solution landed about a week ago, we already removed a little too much information that we should really restore around the manual installation process.

simon66 commented:
Thanks for the feedback @Pomax! I actually like the idea of the "manual installation" part. I noticed that it was there before, but it has since been removed.
I'll update/add the missing portion :)

simon66 commented:
Hi @Pomax. I've added the requested section. I went back to the previous versions of the README.md file and added them in.
Let me know if theres anything that needs to be changed

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,105 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+## 1) Pre-Requisites
+In order for Thimble to be installed correctly, the following dependencies needs to be installed:
 
+- Brackets (Bramble) [[download](https://github.com/mozilla/brackets)]
```
let's remove the "download" link and instead link it to the "Installing Brackets (Bramble)" section

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,105 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
 
-The setup of Thimble can be divided into two distinct sections:
+### Installing Brackets (Bramble)
+* Fork and clone https://github.com/mozilla/brackets
```
I honestly think that the instructions in https://github.com/mozilla/thimble.mozilla.org/pull/1683/files#diff-04c6e90faac2675aa89e2176d2eec7d8L38 (and below) to install Brackets are a bit clearer so I would undo this change. (I still need to update the Wiki with the correct instructions for manual install)

simon66 replied:
I'm a bit confused here. So you want me to revert just the Brackets installation section?

humphd replied:
Ideally you want to clone Brackets with `git clone --recursive ...` to avoid having to also init/update the submodules.

gideonthomas replied:
@simon66 yeah I think the instructions previously for the Brackets installation process (which I linked above and is highlighted in red) might be a bit clearer. That being said, the line about `npm start` can probably be what you have here. Also, I would love to hear your thoughts on which of the two instructions are clearer and why.

gideonthomas commented on README.md:
```diff
@@ -91,4 +143,4 @@ Contact Us
 We're a friendly group, so feel free to chat with us in the "Thimble" channel on Mozilla Chat running on [Mattermost](https://about.mattermost.com). To access Mozilla Chat head to [this link]( http://chat.mozillafoundation.org). Note that you will need to create an account first.
 
-You can also download a mobile or desktop client for Mattermost [here](https://about.mattermost.com/download/#mattermostApps).
+You can also download a mobile or desktop client for Mattermost [here](https://about.mattermost.com/download/#mattermostApps).
```
What change was made here?

simon66 replied:
I actually don't know hahah.
Nothing got changed in this part really. Not too sure why github thinks I changed it haha.

gideonthomas replied:
Alright, not a problem!

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,105 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+This will take less time to setup compared to the `vagrant up` command.
 
-The first step is to fork and clone Thimble and navigate to the cloned directory in a terminal shell.
+### Manual Installation (Optional)
```
I don't think we should add all the instructions here in this readme since it's becoming a bit too long. I would instead link to the Wiki page with all this information: https://github.com/mozilla/thimble.mozilla.org/wiki/Setting-Up-Thimble-without-Virtualization

Pomax replied:
@gideonthomas Then we disagree. Remember that this README is also for tomorrow's user, including people well version in development, who can't use vagrant due to Hyper-V requirements on windows. They should not have to hunt for instructions, and having a "more detailed section" at the bottom of the immediate project documentation is quite common for larger projects (Which Thimble definitely falls under). I would very much advocate we keep the manual installation instructions immediately available rather than locked away in a "you need to go back online" (the README.md downloads with thimble as part of a `git clone` operation - the wiki most definitely does not)

gideonthomas replied:
Good point, I don't think I considered the case for the Hyper-V pitfall with Windows. Alright, I think it's fine to leave these in here for now specifically for that case until we figure out a solution that fixes the virtualization problem on Windows entirely.

I would however make some changes:
- Below "Automated Installation" add the following copy `**Note:** If you aren't able to properly run virtualization software on your machine (for e.g. some versions of Windows only allow one virtualization client to run at a time and if that isn't VirtualBox, you can't run the required VirtualBox as well. This is often a problem if you have docker installed on Windows) or are trying to host Thimble on your own, refer to the [Manual Installation](#manual-installation) instructions instead.`
- Remove "Optional"
- Stick to the instructions from the Wiki here: https://github.com/mozilla/thimble.mozilla.org/pull/1683/files#diff-04c6e90faac2675aa89e2176d2eec7d8L38 since they are a bit clearer and more accurate

simon66 commented:
Thanks for the feedback :)
Going to make these changes asap!

humphd commented on README.md:
```diff
@@ -19,53 +19,105 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
-### Editor
+## 2) Thimble Installation
+### Automated Installation (Preferred Method)
+After forking and cloning Thimble, installing it is quite simple. All we need to do is run the following command inside the cloned directory and let vagrant do the heavy lifting for us:
```
Vagrant (capital 'V').

humphd commented:
@simon66 thanks for jumping in to fix docs.  This is great initiative.

gideonthomas commented:
@ryanwarsaw I think the changes you commented on were changes made based on the manual steps that were in the README before. In [this comment](https://github.com/mozilla/thimble.mozilla.org/pull/1683#discussion_r98940626) I recommended that @simon66 use the instructions from the Wiki instead which should probably address all your comments (if not, please let me know so that I can update them).

ryanwarsaw commented:
@gideonthomas Just wanted to make sure these issues were documented 👍

simon66 commented:
I just added/changed the README based on the comments. Hopefully I didn't miss anything. Please let me know if I did :)

Here are my thoughts on the new README:
I think the new file is quite long (And maybe a bit confusing). It would probably be a good idea to put the Manual Installation portion on a new README and link it to the current one.
What do you think?

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+All three services along with Thimble are bundled together using Git subtrees and can either be run together using Vagrant, or [manually](#1b-manual-installation).
 
-### Editor
+# 1) Installing Thimble and Services
```
General comment about headings - let's remove the numbering from all headings since the heading level generally indicates the list level as well

gideonthomas replied:
This heading specifically - let's remove it since it doesn't really add any context that's different than what comes before it. Also, leave all the other heading levels as they are (for e.g. the heading below, don't change it from ## to # to compensate for removing this heading)

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+**Note:** If you aren't able to properly run virtualization software on your machine (for e.g. some versions of Windows only allow one virtualization client to run at a time and if that isn't VirtualBox, you can't run the required VirtualBox as well. This is often a problem if you have docker installed on Windows) or are trying to host Thimble on your own, refer to the [Manual Installation](#1b-m ...
 
-To fully use Thimble locally, you need to first setup Brackets locally first. This can be done by following the steps outlined below:
+### Pre-Requisites for Automated Installation
```
Pre-requisites should be Prerequisites

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
 
-To fully use Thimble locally, you need to first setup Brackets locally first. This can be done by following the steps outlined below:
+### Pre-Requisites for Automated Installation
+In order for Thimble to be installed correctly, the following dependencies needs to be installed in order:
```
"following dependencies needs" should be changed to "following dependencies need"

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
-This process can take a while depending on your internet connection speed as it needs to download all dependencies. Once you see a log that says `Express server listening on http://localhost:3500`, you can access Thimble on [http://localhost:3500](http://localhost:3500).
 
-You can now make changes to the Thimble source code on your system and they should be automatically reflected on [http://localhost:3500](http://localhost:3500).
+Depending on your internet connection speed, this process can take a while (Since it needs to download all dependencies not listed above).
```
lowercase "s" in Since

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+This will take less time to setup compared to the `vagrant up` command.
+
+##1b) Manual Installation
+You can also setup Thimble and it's needed components outside Vagrant and Virtualbox. This might be needed if you want to:
```
"it's" should be "its"

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+- Host your own instance of Thimble
+- Cannot run virtualization software on your computer
+
+### Pre-Requisites for Manual Installation
```
Prerequisites

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+- Cannot run virtualization software on your computer
+
+### Pre-Requisites for Manual Installation
+In order for Thimble to be installed correctly, the following dependencies needs to be installed:
```
"needs" should be "need"

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+In order for Thimble to be installed correctly, the following dependencies needs to be installed:
+
+- Node.js 4.x or above (see note below)
+  - * **Note:** The login.webmaker.org dependency needs a node version of 4.x only while all the other dependencies work with a node version of 4.x and above. We suggest installing [NVM](https://github.com/creationix/nvm) to allow the use of multiple versions of node.
```
there's an extra asterisk here

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+* Run ``cp env.dist .env`` to create an environment file
+* Run ``npm install`` to install dependencies
+* Run ``npm start`` to start the server
+* Once everything is ready and running, Thimble will be available at [http://localhost:3500/](http://localhost:3500/)
```
let's move this line after the `publish.webmaker.org` section as a general statement.

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+* Clone https://github.com/mozilla/publish.webmaker.org
+* Run ``npm install`` to install dependencies
+* Run ``npm run env``
+* Run ``npm install knex -g`` to install knex
```
Let's remove this since this is actually not required

Pomax commented:
Like I said before - long README.md files are perfectly fine for complex projects, and with five services all running at the same time Thimble absolutely qualifies as one. Cloning the repo only gives you the README.md, not any wikis, so keeping all the information a dev needs to get things up and running without needing to load external webpages is a good thing.

Documentation length is never the issue: clarity is. If the REAMDE.md makes sense to someone trying to find the information they need, then length is irrelevant. If, however, there is so much text in such a disorganized fashion that people can't find what they need to know, then even a short README.md can be useless.

gideonthomas commented on README.md:
```diff
@@ -19,53 +19,161 @@ how it works in [this blog post](http://blog.humphd.org/thimble-and-bramble/).
+  * This lets you authenticate your account without needing email
+* Go back to Thimble and Log In with your new account
+
+# 2) Running the parts
```
I would delete this whole section because it seems redundant once you make the changes I suggest above

simon66 commented:
@gideonthomas Done 👍 
Let me know if there are any other changes that needs to be made :)

gideonthomas commented:
@simon66 thank you for taking initiative on this and patiently going through the multiple revisions of this document. I think it looks good to merge now and we will follow up with the manual install steps.

simon66 commented:
Thanks everyone for the positive feedback :) Glad I was able to help.
The install instruction is much better now 👍

Pomax commented:
@gideonthomas of course the best role model behaviour here would be to then say "and we've filed the follow-up issue, it's #......, you're welcome to help us out on that one of course" =)
